# 第一次作業

In [38]:
# 13.34, 13.38, 13.114, 13.117

In [39]:
import pandas as pd
import numpy as np
import scipy.stats as stats
import statsmodels.stats.api as sms
from matplotlib import pyplot as plt
%matplotlib inline
# 設定圖形大小; DPI越大圖越大
plt.rcParams["figure.dpi"] = 130
import seaborn as sns
import sys
import math

<div style="background-color:#D9D9D9;"><font size="5"> 13.34</font></div>

### H0 : var1/var2 = 1
### H1 : var1/var2 ≠ 1

In [40]:
df1 = pd.read_excel('Xr13-34.xlsx')
display(df1.head())
display(df1.tail())
# 轉換成numpy array
Fixed = df1['Fixed'].values
print("Fixed = \n", Fixed)
Commission = df1['Commission'].values
print("Commission = \n", Commission)

Fixed  Commission
0  75558       48705
1  73974       51927
2  68985       56733
3  60239       52997
4  67320       60655

Fixed  Commission
85  49302       58990
86  47334       64509
87  37480       66350
88  61667       56519
89  37098       60039

Fixed = 
 [75558 73974 68985 60239 67320 59924 55618 69653 57640 52603 43686 71253
 71798 57285 54646 54084 55390 45958 75945 47700 47886 56060 53082 47737
 53166 69030 57949 57182 58557 66434 80682 60464 69153 56834 71307 64615
 59779 60874 62703 68311 78605 63332 59328 54840 66256 54615 63535 53124
 55602 55620 48660 49477 61384 56570 47469 55684 60598 49468 62897 72720
 43287 59496 68229 63594 74114 79747 28115 71206 74003 50013 71735 80088
 66187 64392 60750 57071 63557 41185 60070 64852 76579 53649 74806 67698
 63907 49302 47334 37480 61667 37098]
Commission = 
 [48705 51927 56733 52997 60655 61727 59454 62862 59761 75166 69110 72217
 61031 71358 51593 61980 45447 76879 63941 81009 71450 80858 71611 59454
 73211 70225 59719 69578 64372 58567 61718 45120 93770 62140 63285 75151
 47151 76937 72522 72531 53816 50185 69892 63265 73362 69218 65051 61045
 60111 80348 52766 45823 58092 66867 81402 65955 59140 55590 91441 56617
 45831 64169 51746 49121 54199 77570 54961 63142 53824 79837 

In [41]:
df_num = Fixed.shape[0] - 1
df_denom = Commission.shape[0] - 1
print("df_num = ", df_num)
print("df_denom = ", df_denom)
var_ratio = stats.describe(Fixed).variance / stats.describe(Commission).variance
print(f"var_ratio (F value) = {var_ratio:.4f}")
ptmp = stats.f.cdf(var_ratio, df_num, df_denom)
p_value = ptmp * 2
print(f"p_value = {p_value:.4f}")

df_num =  89
df_denom =  89
var_ratio (F value) = 0.9541
p_value = 0.8252


The p-value for equal-variance test is 0.8252, which is larger than the required 0.05 significance level, there is no sufficient evidence to reject the null hypothesis. As a result, we assume that the two populations have equal variance and apply equal-variance t-test on μ1-μ2 = 0 <br>
### H0:μ1-μ2>=0 , 
### H1:μ1-μ2<0



In [42]:
# t-test using sms.CompareMeans
cm = sms.CompareMeans(sms.DescrStatsW(Fixed), sms.DescrStatsW(Commission))
# usevarstr, ‘pooled’ or ‘unequal’
cm.summary(usevar='pooled', alpha=0.05)

<class 'statsmodels.iolib.table.SimpleTable'>

Based on the above result, the one-sided p-value should be 0.038/2 = 0.019, and that is lesser than our significant level. So we have enough evidence at a significance level of 0.05 to reject the null hypothesis in favor of the alternative one and conclude that the commission salesperson outperforms the fixed-salary salesperson.

<div style="background-color:#D9D9D9;"><font size="5"> 13.38</font></div>

### H0 : var1/var2 = 1
### H1 : var1/var2 ≠ 1

In [43]:
df1 = pd.read_excel('Xr13-38.xlsx')
display(df1.head())
display(df1.tail())
# 轉換成numpy array
BB = df1['Big Bucket'].values
print("Big = \n", BB)
MB = df1['Medium Bucket'].values
print("Medium = \n", MB)

Big Bucket  Medium Bucket
0         133             50
1         106             63
2         104             56
3          85             37
4          63             63

Big Bucket  Medium Bucket
43          86             51
44          96             59
45         116             66
46          93             73
47          85             70

Big = 
 [133 106 104  85  63 100  99  91  81  94  91 135  76 104  87 107  98  90
  90  84  91  96  92  89  92 118 114  76 104  89  95 111  62  69 123  95
  75  97  67  98  80  82  94  86  96 116  93  85]
Medium = 
 [50 63 56 37 63 64 52 73 60 72 58 67 79 55 58 45 65 66 69 59 64 54 59 72
 72 60 66 60 59 55 50 44 71 69 64 66 44 63 66 60 52 67 73 51 59 66 73 70]


In [44]:
df_num = BB.shape[0] - 1
df_denom = MB.shape[0] - 1
print("df_num = ", df_num)
print("df_denom = ", df_denom)
var_ratio = stats.describe(BB).variance / stats.describe(MB).variance
print(f"var_ratio (F value) = {var_ratio:.4f}")
ptmp = stats.f.cdf(var_ratio, df_num, df_denom)
p_value = (1-ptmp) * 2
print(f"p_value = {p_value:.4f}")

df_num =  47
df_denom =  47
var_ratio (F value) = 3.1431
p_value = 0.0001


The p-value for equal-variance test is 0.0001, which is smaller than the required 0.05 significance level, 
there is sufficient evidence to reject the null hypothesis. As a result, we assume that the two populations 
don't have equal variance and apply nonequal-variance t-test on μ1-μ2 = 0 <br>
### H0:μ1-μ2<=0 , 
### H1:μ1-μ2>0

In [45]:
# t-test using sms.CompareMeans
cm = sms.CompareMeans(sms.DescrStatsW(BB), sms.DescrStatsW(MB))
# usevarstr, ‘pooled’ or ‘unequal’
cm.summary(usevar='unequal', alpha=0.05)

<class 'statsmodels.iolib.table.SimpleTable'>

Based on the above result, the one-sided p-value should be 0, and that is lesser than our 
significant level. So we have enough evidence at a significance level of 0.05 to reject the null hypothesis 
in favor of the alternative one and conclude that the larger the bucket the more people will eat.

<div style="background-color:#D9D9D9;"><font size="5"> 13.114</font></div>

### H0 : var1/var2 = 1
### H1 : var1/var2 ≠ 1

In [46]:
df1 = pd.read_excel('Xr13-114.xlsx')
display(df1.head())
display(df1.tail())
# 轉換成numpy array
M1 = df1['Machine 1'].values
M1 = M1[~np.isnan(M1)]
print("M1 = \n", M1)
M2 = df1['Machine 2'].values
print("M2 = \n", M2)

Machine 1  Machine 2
0      0.998      1.003
1      0.997      1.004
2      1.003      0.997
3      1.000      0.996
4      0.999      0.999

Machine 1  Machine 2
6       0.998      1.000
7       1.003      1.005
8       1.004      1.002
9       1.000      1.004
10        NaN      0.996

M1 = 
 [0.998 0.997 1.003 1.    0.999 1.    0.998 1.003 1.004 1.   ]
M2 = 
 [1.003 1.004 0.997 0.996 0.999 1.003 1.    1.005 1.002 1.004 0.996]


In [47]:
df_num = M1.shape[0] - 1
df_denom = M2.shape[0] - 1
print("df_num = ", df_num)
print("df_denom = ", df_denom)
var_ratio = stats.describe(M1).variance / stats.describe(M2).variance
print(f"var_ratio (F value) = {var_ratio:.4f}")
ptmp = stats.f.cdf(var_ratio, df_num, df_denom)
p_value = (ptmp) * 2
print(f"p_value = {p_value:.4f}")

df_num =  9
df_denom =  10
var_ratio (F value) = 0.5045
p_value = 0.3179


The p-value for equal-variance test is 0.3179, which is larger than the required 0.05 significance level, 
there is no sufficient evidence to reject the null hypothesis. As a result, we assume that the two populations 
have equal variance, the manager can't infer that the two machines differ in their consistency of fills.

<div style="background-color:#D9D9D9;"><font size="5"> 13.117</font></div>

### H0 : var1/var2 = 1
### H1 : var1/var2 ≠ 1

In [48]:
df1 = pd.read_excel('Xr13-117.xlsx')
display(df1.head())
display(df1.tail())
# 轉換成numpy array
W1 = df1['Week 1'].values
print("W1 = \n", W1)
W2 = df1['Week 2'].values
print("W2 = \n", W2)

Week 1  Week 2
0      69      75
1      77      69
2      74      73
3      67      72
4      81      71

Week 1  Week 2
95      67      76
96      84      69
97      71      68
98      73      74
99      73      64

W1 = 
 [69 77 74 67 81 67 72 68 68 68 67 67 74 64 66 76 71 67 67 67 79 77 69 68
 74 71 71 75 68 70 71 75 73 72 72 68 68 73 71 82 73 69 68 71 68 69 64 71
 65 70 77 75 72 71 65 63 65 71 76 65 68 74 72 72 75 71 67 78 75 68 75 66
 72 74 73 75 74 69 74 68 70 64 68 64 64 77 69 76 78 63 74 68 63 70 68 67
 84 71 73 73]
W2 = 
 [75 69 73 72 71 69 73 75 70 72 76 65 65 65 70 70 73 72 71 71 67 68 68 74
 72 79 66 71 67 73 71 71 72 68 78 65 70 67 72 71 71 66 64 79 70 71 74 71
 70 73 67 70 71 65 67 67 75 72 68 75 69 72 72 78 77 73 73 69 71 69 77 77
 72 68 72 74 72 75 69 71 71 68 71 76 68 69 69 72 63 74 77 68 71 74 76 76
 69 68 74 64]


In [49]:
df_num = W1.shape[0] - 1
df_denom = W2.shape[0] - 1
print("df_num = ", df_num)
print("df_denom = ", df_denom)
var_ratio = stats.describe(W1).variance / stats.describe(W2).variance
print(f"var_ratio (F value) = {var_ratio:.4f}")
ptmp = stats.f.cdf(var_ratio, df_num, df_denom)
p_value = (1-ptmp) * 2
print(f"p_value = {p_value:.4f}")

df_num =  99
df_denom =  99
var_ratio (F value) = 1.5261
p_value = 0.0366


The p-value for equal-variance test is 0.0366, which is smaller than the required 0.05 significance level, so 
there is sufficient evidence to reject the null hypothesis. As a result, we assume that the two populations 
don't have equal variance, also because the speed variance of the first week is larger than second week.we can infer that limiting the minimum and maximum speeds reduces the variation in speeds.